### PR TITLE
6519: adds trigger events to known names

### LINKED ISSUE #6519: Modify event name collector to look at custom event trigger names configured for starter bricks
### Context
https://github.com/pixiebrix/pixiebrix-extension/pull/6288#pullrequestreview-1585217018

### PR DESCRIPTION
**Pull request by grahamlangford**
## What does this PR do?

- Closes #6519 

## Demo
https://www.loom.com/share/e0ae7cbd42a74d828071d58961d8916a

## Checklist

- [x] Add tests
- [x] Designate a primary reviewer @twschiller

### CODE REVIEW

File: src/types/starterBrickTypes.ts [hunk 1]
```diff
@@ -24,6 +24,7 @@ import { type Brick } from "@/types/brickTypes";
 import { type Reader } from "@/types/bricks/readerTypes";
 import { type Metadata } from "@/types/registryTypes";
 import { type UnknownObject } from "@/types/objectTypes";
+import { type StarterBrickType } from "@/starterBricks/types";
```
Comment (twschiller) [inline src/types/starterBrickTypes.ts:27]: We may want to/need to move the StarterBrickType type to this file. The idea behind the top-level type's folder to hold types that can be safely imported from anywhere without introducing circular module dependencies

File: src/analysis/analysisVisitors/eventNameAnalysis/collectEventNamesVisitor.ts [hunk 2]
```diff
@@ -40,15 +49,29 @@ export type EventNameAnalysisResult = {
...
 
   public get result(): EventNameAnalysisResult {
     return {
       knownNames: [...this._eventNames],
+      triggerNames: [...this._triggerNames],
       hasDynamicEventName: this._hasDynamicEventName,
     };
   }
 
+  private visitExtensionPoint(
```
Comment (twschiller) [inline src/analysis/analysisVisitors/eventNameAnalysis/collectEventNamesVisitor.ts:63]: 👍

File: src/analysis/analysisVisitors/eventNameAnalysis/collectEventNamesVisitor.ts [hunk 3]
```diff
@@ -21,13 +21,22 @@ import PipelineVisitor, {
...
+} from "@/pageEditor/starterBricks/formStateTypes";
 
 export type EventNameAnalysisResult = {
   /**
    * Statically known event names in the FormState.
    */
   knownNames: string[];
+  /**
+   * Event names that are used by a trigger starter brick.
+   * Kept separate for analysis. Combined with knownNames for runtime.
```
Comment (twschiller) [inline src/analysis/analysisVisitors/eventNameAnalysis/collectEventNamesVisitor.ts:36]: NIT: I'm not sure I follow this comment. Why do they need to be kept separate for the analysis? Also, IIRC runtime doesn't use this analysis to run bricks. Is the comment referring to which event name are suggested for the Emit Event brick?

File: src/analysis/analysisVisitors/eventNameAnalysis/checkEventNamesAnalysis.ts [hunk 4]
```diff
@@ -43,7 +43,14 @@ class CheckEventNamesAnalysis extends AnalysisVisitorABC {
   }
 
   get knownEventNames(): string[] {
```
Comment (twschiller) [inline src/analysis/analysisVisitors/eventNameAnalysis/checkEventNamesAnalysis.ts:45]: The "knownEventNames" vs. "knownNames" distinction is subtle. We probably should improve the naming of `knownNames`, e.g., `knownEmittedNames`?

**Review by twschiller — APPROVED** [review]
Functionality looks correct, left some comments on clarifying name/comments on the relationship between the two fields in the analysis result

File: src/analysis/analysisVisitors/eventNameAnalysis/checkEventNamesAnalysis.ts (hunk 4 above)
Comment (grahamlangford) [reply src/analysis/analysisVisitors/eventNameAnalysis/checkEventNamesAnalysis.ts:45]: That's a good point. I'll update the name.

File: src/analysis/analysisVisitors/eventNameAnalysis/collectEventNamesVisitor.ts (hunk 3 above)
Comment (grahamlangford) [reply src/analysis/analysisVisitors/eventNameAnalysis/collectEventNamesVisitor.ts:36]: That is what I meant. I updated the comment to be more clear.

File: src/types/starterBrickTypes.ts (hunk 1 above)
Comment (grahamlangford) [reply src/types/starterBrickTypes.ts:27]: I didn't see any circular dependency warning, but I'll move it anyway.

File: src/analysis/analysisVisitors/eventNameAnalysis/collectEventNamesVisitor.ts [hunk 5]
```diff
@@ -40,15 +50,29 @@ export type EventNameAnalysisResult = {
...
   public get result(): EventNameAnalysisResult {
     return {
-      knownNames: [...this._eventNames],
+      knownEmittedNames: [...this._eventNames],
+      knownTriggerNames: [...this._triggerNames],
       hasDynamicEventName: this._hasDynamicEventName,
     };
   }
 
+  private visitStarterBrick(
```
Comment (grahamlangford) [inline src/analysis/analysisVisitors/eventNameAnalysis/collectEventNamesVisitor.ts:64]: Calling out that I changed the name of this method from `visitExtensionPoint`

File: src/analysis/analysisVisitors/eventNameAnalysis/collectEventNamesVisitor.ts [hunk 6]
```diff
@@ -21,13 +21,23 @@ import PipelineVisitor, {
...
    */
-  knownNames: string[];
+  knownEmittedNames: string[];
+  /**
+   * Event names that are used by a trigger starter brick.
+   * Kept separate to preserve warnings in the trigger starter brick when the event name is not emitted from another brick.
+   * Combined with knownEmittedNames to get all known event names to set in redux.
+   * @since 1.8.2
+   */
+  knownTriggerNames: string[];
```
Comment (grahamlangford) [inline src/analysis/analysisVisitors/eventNameAnalysis/collectEventNamesVisitor.ts:40]: I'm still unsure if I'm happy with this description.

Comment (twschiller) [reply src/analysis/analysisVisitors/eventNameAnalysis/collectEventNamesVisitor.ts:40]: Definitely good enough to provide context

**Pull request merged**